lib/phoenix/endpoint.ex

```diff
@@ -240,30 +240,81 @@ defmodule Phoenix.Endpoint do
-       where the instrumentation is being done). `runtime_meta` is a term that
-       is passed on by the caller of the instrumentation. The result of this
-       call is stored and later passed to the after callback.
-    2. The event happens (in this case, the view is rendered).
-    3. The event callback is called again, this time with the following arguments:
+  #### "Before" clause
+
+  When the first argument to an event callback is `:start`, the signature of
+  that callback is:
+
+      event_callback(:start, compile_metadata, runtime_metadata)
+
+  where:
+
+    * `compile_metadata` is a map of compile-time metadata about the environment
+      where `instrument/3` has been called. It contains the module where the
+      instrumentation is happening (under the `:module` key), the file and line
+      (`:file` and `:line`), and the function inside which the instrumentation
+      is happening (under `:function`). This information can be used arbitrarely
+      by the callback.
+    * `runtime_metadata` is a map of runtime data that the instrumentation
+      passes to the callbacks. This can be used for any purposes: for example,
+      when instrumenting the rendering of a view, the name of the view could be
+      passed in these runtime data so that instrumenters know which view is
+      being rendered (`instrument(:view_render, %{view: "index.html"}, fn
+      ...)`).
+
+  #### "After" clause
+
+  When the first argument to an event callback is `:stop`, the signature of that
+  callback is:
 
-           MyInstrumenter.render_view(:stop, time_diff, start_result)
+      event_callback(:stop, time_diff, result_of_before_callback)
 
-       `time_diff` is the time *in microseconds* it took for the event to
-       happen (in this case, the view rendering time). `start_result` is
-       whatever the event callback returned when called with `:start` as the
-       first argument: instrumenters can use this to pass "state" from the
-       before callback to the after callback.
+  where:
+
+    * `time_diff` is an integer representing the time it took to execute the
+      instrumented function **in microseconds**.
+    * `result_of_before_callback` is the return value of the "before" clause of
+      the same `event_callback`. This is a means of passing data from the
+      "before" clause to the "after" clause when instrumenting. For example, an
+      instrumenter can implement custom time measuring with this:
+
+          defmodule MyInstrumenter do
+            def event_callback(:start, _compile, _runtime) do
+              :erlang.monotonic_time(:micro_seconds)
+            end
+
+            def event_callback(:stop, _time_diff, start_time) do
+              stop_time = :erlang.monotonic_time(:micro_seconds)
+              do_something_with_diff(stop_time - start_time)
+            end
+          end
+
+  The return value of each "before" event callback will be stored and passed to
+  the corresponding "after" callback.
```

Comment:
+1 on the same process thing, but I think we always call "after" callbacks? https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/endpoint/instrument.ex#L67-L72